src/getPublicKeys.c

```diff
@@ -0,0 +1,265 @@
+#include "state.h"
+#include "securityPolicy.h"
+#include "uiHelpers.h"
+#include "uiScreens.h"
+#include "getPublicKeys.h"
+
+static int16_t RESPONSE_READY_MAGIC = 23456;
+
+static ins_get_keys_context_t* ctx = &(instructionState.getKeysContext);
+
+static inline void CHECK_STAGE(get_keys_stage_t expected)
+{
+	TRACE("Checking stage... current one is %d, expected %d", ctx->stage, expected);
+	VALIDATE(ctx->stage == expected, ERR_INVALID_STATE);
+}
+
+// read a path from view into ctx->pathSpec
+static void parsePath(read_view_t* view)
+{
+	view_skipBytes(view, bip44_parseFromWire(&ctx->pathSpec, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view)));
+	BIP44_PRINTF(&ctx->pathSpec);
+}
+
+// ============================== derivation and UI state machine for one key ==============================
+
+enum {
+	GET_KEY_UI_STEP_WARNING = 200,
+	GET_KEY_UI_STEP_DISPLAY,
+	GET_KEY_UI_STEP_CONFIRM,
+	GET_KEY_UI_STEP_RESPOND,
+	GET_KEY_UI_STEP_INVALID,
+} ;
+
+static void getPublicKeys_getOneKey_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = getPublicKeys_getOneKey_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+	UI_STEP(GET_KEY_UI_STEP_WARNING) {
+		ui_displayPaginatedText(
+		        "Unusual request",
+		        "Proceed with care",
+		        this_fn
+		);
+	}
+	UI_STEP(GET_KEY_UI_STEP_DISPLAY) {
+		ui_displayAccountScreen("Export public key", &ctx->pathSpec, this_fn);
+	}
+	UI_STEP(GET_KEY_UI_STEP_CONFIRM) {
+		ui_displayPrompt(
+		        "Confirm export",
+		        "public key?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(GET_KEY_UI_STEP_RESPOND) {
+		ASSERT(ctx->responseReadyMagic == RESPONSE_READY_MAGIC);
+
+		io_send_buf(SUCCESS, (uint8_t*) &ctx->extPubKey, SIZEOF(ctx->extPubKey));
+		ctx->responseReadyMagic = 0; // just for safety
+
+		if (ctx->remainingPaths > 0) {
+			ui_displayBusy(); // waiting for another APDU
+		} else {
+			ui_idle(); // we are done, display the main app screen
+		}
+	}
+	UI_STEP_END(GET_KEY_UI_STEP_INVALID);
+}
+
+// derive the key described by ctx->pathSpec and run the ui state machine accordingly
+// if isSingleKeyExport is true, more restrictive security policy is applied
+static void returnOneKey(bool isSingleKeyExport)
```

Comment:
can we replace boolean with enum?

Reply:
Actually, why don't we
1) rename this to displayPromptAndRespond() or something like that?
2) it would make sense to just pass security_policy to this function instead of passing in boolean/enum